### DISCUSSION TITLE
Looking for X11 automation similar to pywinauto

### DISCUSSION THREAD

**Discussion by louiscaron** [pywinauto/pywinauto#1106] [Q&A] [answered]
Hello and thanks for the great work!

In my organisation, I have been using pywinauto to automate some UI procedures that could not be done using the sole command line interface.  In an effort to move away from windows, I am evaluating the effort to port the UI automation script to the linux environment (our tool has a linux version, including the GUI as well).

I saw that you are participating as well to the python-xlib project.  Is this the right framework to drive UI procedures?  I have the feeling that it is more a framework to develop native X11 applications rather than perform tests on an existing application (I could not find anything like find_window etc...).

Do you have any advice for an X11 framework to automate UI procedures?

thanks a lot

**Comment by vasily-v-ryabov** [comment, marked as answer]
Hi @louiscaron we're working on Linux support using ATSPI. You can try Linux version from this branch:
```sh
pip uninstall -y pywinauto
pip install https://github.com/pywinauto/pywinauto/archive/refs/heads/atspi.zip
```
"atspi" is a current active branch which has kind of alpha quality. It is not released yet. Early testing and feedback is welcome!

**Reply by louiscaron** [reply]
Thanks a lot for the quick answer!  It looks like atspi has a very limited interaction capability!  

And this seems to be because of the nature of X11, therefore I assume that there is no framework as good as pywinauto for X11 that would support searching in all the windows present on the screen with filters, retrieve the class of objects and so on.

**Reply by vasily-v-ryabov** [reply]
"atspi" backend in pywinauto is not complete yet (contributions are welcome). Technically you can try to use [pyatspi2](https://github.com/GNOME/pyatspi2). Last time we tried it (~3 ears ago) it had compilation problems and wasn't looking useful. But the situation might be changed.

X11 (python-xlib) provides only access to top-level windows and can't work with widgets on them (buttons, combo boxes etc). ATSPI can access widgets on many apps. But the app should also support ATSPI in its own code. For example, Chrome has good accessibility support on all OSes. Qt apps should be good as well. Maybe GTK.

**Reply by louiscaron** [reply]
Thanks a lot for the reply, I saw that atspi page points to https://gitlab.gnome.org/GNOME/accerciser to test if the targetted application supports atspi, I will start with this.
